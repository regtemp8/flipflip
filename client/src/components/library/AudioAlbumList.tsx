import React, { useEffect, useState } from 'react'
import { cx } from '@emotion/css'

import {
  Card,
  CardContent,
  CardMedia,
  Grid,
  type Theme,
  Tooltip,
  Typography
} from '@mui/material'

import { makeStyles } from 'tss-react/mui'

import AudiotrackIcon from '@mui/icons-material/Audiotrack'

import type Audio from '../../store/audio/Audio'

const useStyles = makeStyles()((theme: Theme) => ({
  emptyMessage: {
    textAlign: 'center',
    marginTop: '25%'
  },
  emptyMessage2: {
    textAlign: 'center'
  },
  media: {
    height: 0,
    paddingTop: '100%' // 16:9  = 56.25%
  },
  mediaIcon: {
    width: '100%',
    height: 'auto'
  },
  underlineTitle: {
    textDecoration: 'underline'
  },
  pointer: {
    cursor: 'pointer'
  },
  root: {
    borderRadius: 1
  },
  cardContent: {
    '&:last-child': {
      paddingBottom: theme.spacing(2)
    }
  },
  artist: {
    '&:hover': {
      textDecoration: 'underline'
    }
  }
}))

export interface AudioAlbumListProps {
  sources: Audio[]
  showHelp: boolean
  onClickAlbum: (album: string) => void
  onClickArtist: (artist: string) => void
}

function AudioAlbumList(props: AudioAlbumListProps) {
  const [albums, setAlbums] = useState<
    Map<string, { artist: string; thumb?: string; count: number }>
  >(new Map())
  const [hover, setHover] = useState<string>()

  useEffect(() => {
    setAlbums(getAlbums(props.sources))
  }, [props.sources])

  const nop = () => {}

  const onMouseEnter = (album: string) => {
    setHover(album)
  }

  const onMouseLeave = () => {
    setHover(undefined)
  }

  const getAlbums = (
    sources: Audio[]
  ): Map<string, { artist: string; thumb?: string; count: number }> => {
    const va = 'Various Artists'
    const albumMap = new Map<
      string,
      { artist: string; thumb?: string; count: number }
    >()
    const songs = Array.from(sources).sort((a, b) => {
      const albumA = a.album as string
      const albumB = b.album as string
      if (albumA > albumB) {
        return 1
      } else if (albumA < albumB) {
        return -1
      } else {
        const trackNumA = a.trackNum as number
        const trackNumB = b.trackNum as number
        if (trackNumA > trackNumB) {
          return 1
        } else if (trackNumA < trackNumB) {
          return -1
        } else {
          const nameA = a.name as string
          const nameB = b.name as string
          const reA = /^(A\s|a\s|The\s|the\s)/g
          const valueA = nameA.replace(reA, '')
          const valueB = nameB.replace(reA, '')
          return valueA.localeCompare(valueB, 'en', { numeric: true })
        }
      }
    })
    for (const song of songs) {
      if (
        song.album &&
        (!albumMap.has(song.album) ||
          !albumMap.get(song.album)?.thumb ||
          (albumMap.get(song.album)?.artist !== song.artist &&
            albumMap.get(song.album)?.artist !== va))
      ) {
        if (
          albumMap.has(song.album) &&
          albumMap.get(song.album)?.artist !== song.artist
        ) {
          albumMap.set(song.album, { artist: va, thumb: song.thumb, count: 0 })
        } else {
          albumMap.set(song.album, {
            artist: song.artist as string,
            thumb: song.thumb,
            count: 0
          })
        }
      }
    }
    for (const song of songs) {
      if (song.album) {
        const album = albumMap.get(song.album)
        albumMap.set(song.album, {
          artist: album?.artist as string,
          thumb: album?.thumb,
          count: (album?.count as number) + 1
        })
      }
    }
    return albumMap
  }

  const { classes } = useStyles()
  if (albums.size === 0) {
    return (
      <React.Fragment>
        <Typography
          component="h1"
          variant="h3"
          color="inherit"
          noWrap
          className={classes.emptyMessage}
        >
          乁( ◔ ౪◔)「
        </Typography>
        <Typography
          component="h1"
          variant="h4"
          color="inherit"
          noWrap
          className={classes.emptyMessage2}
        >
          Nothing here
        </Typography>
        {props.showHelp && (
          <Typography
            component="h1"
            variant="h6"
            color="inherit"
            noWrap
            className={classes.emptyMessage2}
          >
            Add tracks by going to the "Songs" tab and clicking the +
          </Typography>
        )}
      </React.Fragment>
    )
  }

  const va = 'Various Artists'
  const albumsArray = Array.from(albums.entries())
  return (
    <Grid container spacing={2}>
      {albumsArray.map(([a, data]) => {
        let thumb: string | undefined = data.thumb
        if (thumb) thumb = thumb.replace(/\\/g, '/')
        const artist = data.artist
        const count = data.count
        return (
          <Grid
            key={a}
            item
            xs={6}
            sm={4}
            md={3}
            lg={2}
            className={classes.pointer}
            onClick={() => props.onClickAlbum(a)}
            onMouseEnter={() => onMouseEnter(a)}
            onMouseLeave={onMouseLeave}
          >
            <Card classes={{ root: classes.root }}>
              {thumb && (
                <CardMedia className={classes.media} image={thumb} title={a} />
              )}
              {!thumb && <AudiotrackIcon className={classes.mediaIcon} />}
              <CardContent classes={{ root: classes.cardContent }}>
                <Tooltip disableInteractive title={a} enterDelay={800}>
                  <Typography
                    className={cx(hover === a && classes.underlineTitle)}
                    noWrap
                    variant="body1"
                  >
                    {a}
                  </Typography>
                </Tooltip>
                <Typography
                  id={'artist-link'}
                  noWrap
                  onClick={() => {
                    if (artist !== va) {
                      props.onClickArtist(artist)
                    }
                  }}
                  className={cx(artist !== va && classes.artist)}
                  color="textSecondary"
                  variant="body2"
                >
                  {artist}
                </Typography>
                <Typography noWrap color="textSecondary" variant="body2">
                  {count} {count === 1 ? 'song' : 'songs'}
                </Typography>
              </CardContent>
            </Card>
          </Grid>
        )
      })}
    </Grid>
  )
}

;(AudioAlbumList as any).displayName = 'AudioAlbumList'
export default AudioAlbumList
