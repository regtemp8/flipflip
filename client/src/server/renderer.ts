import createCache, { type EmotionCache } from '@emotion/cache'
import { configureStore } from '@reduxjs/toolkit'
import { AppStorage, SystemConstants } from 'flipflip-common'
import createAppReducer from '../store/app/slice'
import createAudioReducer from '../store/audio/slice'
import createCaptionScriptReducer from '../store/captionScript/slice'
import createClipReducer from '../store/clip/slice'
import createLibrarySourceReducer from '../store/librarySource/slice'
import createPlaylistReducer from '../store/playlist/slice'
import createSceneReducer from '../store/scene/slice'
import createSceneGroupReducer from '../store/sceneGroup/slice'
import createTagReducer from '../store/tag/slice'
import playersSlice from '../store/player/slice'
import videoClipperSlice from '../store/videoClipper/slice'
import scenePickerSlice from '../store/scenePicker/slice'
import captionScriptorSlice from '../store/captionScriptor/slice'
import sceneDetailSlice from '../store/sceneDetail/slice'
import audioLibrarySlice from '../store/audioLibrary/slice'
import sourceScraperSlice from '../store/sourceScraper/slice'
import createConstantsReducer from '../store/constants/slice'
import { fromAppStorage } from '../store/app/convert'
import createDisplayReducer from '../store/display/slice'
import createDisplayViewReducer from '../store/displayView/slice'
import createScenePlaylistItemReducer from '../store/scenePlaylistItem/slice'
import createDisplayPlaylistItemReducer from '../store/displayPlaylistItem/slice'

export function createEmotionCache(): EmotionCache {
  return createCache({ key: 'css' })
}

export function createReduxStore(
  appStorage?: AppStorage,
  constants?: SystemConstants
) {
  const state = appStorage != null ? fromAppStorage(appStorage) : undefined
  return configureStore({
    reducer: {
      app: createAppReducer(state?.app),
      audio: createAudioReducer(state?.audio),
      captionScript: createCaptionScriptReducer(state?.captionScript),
      clip: createClipReducer(state?.clip),
      librarySource: createLibrarySourceReducer(state?.librarySource),
      playlist: createPlaylistReducer(state?.playlist),
      displayPlaylistItem: createDisplayPlaylistItemReducer(
        state?.displayPlaylistItem
      ),
      scenePlaylistItem: createScenePlaylistItemReducer(
        state?.scenePlaylistItem
      ),
      scene: createSceneReducer(state?.scene),
      sceneGroup: createSceneGroupReducer(state?.sceneGroup),
      display: createDisplayReducer(state?.display),
      displayView: createDisplayViewReducer(state?.displayView),
      tag: createTagReducer(state?.tag),
      constants: createConstantsReducer(constants),
      // components
      players: playersSlice,
      videoClipper: videoClipperSlice,
      scenePicker: scenePickerSlice,
      captionScriptor: captionScriptorSlice,
      sceneDetail: sceneDetailSlice,
      audioLibrary: audioLibrarySlice,
      sourceScraper: sourceScraperSlice
    }
  })
}
