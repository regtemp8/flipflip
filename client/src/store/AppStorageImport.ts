import { initialApp } from "./app/data/App";
import { initialAudioState } from "./audio/slice";
import { initialCaptionScriptState } from "./captionScript/slice";
import { initialClipState } from "./clip/slice";
import { initialLibrarySourceState } from "./librarySource/slice";
import { initialOverlayState } from "./overlay/slice";
import { initialPlaylistState } from "./playlist/slice";
import { initialSceneState } from "./scene/slice";
import { initialSceneGridState } from "./sceneGrid/slice";
import { initialSceneGroupState } from "./sceneGroup/slice";
import { initialTagState } from "./tag/slice";
import { EntryState } from "./EntryState";
import type Tag from "./tag/Tag";
import type SceneGroup from "./sceneGroup/SceneGroup";
import type SceneGrid from "./sceneGrid/SceneGrid";
import type Scene from "./scene/Scene";
import type Playlist from "./playlist/Playlist";
import type Overlay from "./overlay/Overlay";
import type LibrarySource from "./librarySource/LibrarySource";
import type Clip from "./clip/Clip";
import type CaptionScript from "./captionScript/CaptionScript";
import type Audio from "./audio/Audio";
import type App from "./app/data/App";

export interface AppStorageImport {
  app: App;
  audio: EntryState<Audio>;
  captionScript: EntryState<CaptionScript>;
  clip: EntryState<Clip>;
  librarySource: EntryState<LibrarySource>;
  overlay: EntryState<Overlay>;
  playlist: EntryState<Playlist>;
  scene: EntryState<Scene>;
  sceneGrid: EntryState<SceneGrid>;
  sceneGroup: EntryState<SceneGroup>;
  tag: EntryState<Tag>;
}

export const initialAppStorageImport: AppStorageImport = {
  app: initialApp,
  audio: initialAudioState,
  captionScript: initialCaptionScriptState,
  clip: initialClipState,
  librarySource: initialLibrarySourceState,
  overlay: initialOverlayState,
  playlist: initialPlaylistState,
  scene: initialSceneState,
  sceneGrid: initialSceneGridState,
  sceneGroup: initialSceneGroupState,
  tag: initialTagState,
}