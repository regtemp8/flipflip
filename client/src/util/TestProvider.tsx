import * as React from 'react'
import { Provider } from 'react-redux'
import { createTheme } from '@mui/material'
import { StyledEngineProvider, ThemeProvider } from '@mui/material/styles'
import defaultTheme from 'flipflip-common'

export default class TestProvider extends React.Component {
  readonly props: {
    children: any
    store: any
  }

  disableTransitions = {
    defaultProps: {
      disablePortal: true,
      hideBackdrop: true,
      TransitionComponent: ({ children }) => children
    }
  }

  render() {
    const theme = createTheme({
      ...defaultTheme,
      // disable Transitions when runing tests, throws errors otherwise.
      components: {
        MuiMenu: this.disableTransitions,
        MuiPopover: this.disableTransitions,
        MuiDialog: this.disableTransitions,
        MuiModal: this.disableTransitions
      }
    } as any)
    return (
      <Provider store={this.props.store}>
        <StyledEngineProvider injectFirst>
          <ThemeProvider theme={theme}>{this.props.children}</ThemeProvider>
        </StyledEngineProvider>
      </Provider>
    )
  }
}
