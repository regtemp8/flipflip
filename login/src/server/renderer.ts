import createCache, { type EmotionCache } from "@emotion/cache";

export function createEmotionCache(nonce?: string): EmotionCache {
  return createCache({ key: "css", nonce });
}
