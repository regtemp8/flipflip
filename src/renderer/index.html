<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <title>FlipFlip</title>
    <!-- Responsive meta tag -- ensures proper rendering and touch zooming for all devices -->
    <meta name="viewport" content="minimum-scale=1, initial-scale=1, width=device-width, shrink-to-fit=no" />
    <!-- Roboto Font -->
    <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700&display=swap" />
    <!-- Font Icons -->
    <link rel="stylesheet" href="https://fonts.googleapis.com/icon?family=Material+Icons" />

    <meta
      http-equiv="Content-Security-Policy"
      content="frame-src https://www.imagefap.com http://www.imagefap.com https://hypno.nimja.com/ http://hypno.nimja.com/; font-src 'self' https://fonts.gstatic.com; style-src 'self' https://fonts.googleapis.com 'unsafe-inline'"
    />
  </head>

  <body>
    <div id="app" class="app"></div>
  </body>
</html>
