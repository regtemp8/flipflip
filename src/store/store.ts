import { configureStore } from '@reduxjs/toolkit'
import appSlice from './app/slice'
import audioSlice from './audio/slice'
import constantsSlice from './constants/slice'
import playersSlice from './player/slice'
import captionScriptSlice from './captionScript/slice'
import clipSlice from './clip/slice'
import librarySourceSlice from './librarySource/slice'
import overlaySlice from './overlay/slice'
import playlistSlice from './playlist/slice'
import sceneSlice from './scene/slice'
import sceneGridSlice from './sceneGrid/slice'
import sceneGroupSlice from './sceneGroup/slice'
import tagSlice from './tag/slice'
import videoClipperSlice from './videoClipper/slice'
import scenePickerSlice from './scenePicker/slice'
import captionScriptorSlice from './captionScriptor/slice'
import sceneDetailSlice from './sceneDetail/slice'
import audioLibrarySlice from './audioLibrary/slice'
import sourceScraperSlice from './sourceScraper/slice'

const store = configureStore({
  reducer: {
    app: appSlice,
    audio: audioSlice,
    captionScript: captionScriptSlice,
    clip: clipSlice,
    constants: constantsSlice,
    librarySource: librarySourceSlice,
    overlay: overlaySlice,
    playlist: playlistSlice,
    scene: sceneSlice,
    sceneGrid: sceneGridSlice,
    sceneGroup: sceneGroupSlice,
    tag: tagSlice,
    // components
    players: playersSlice,
    videoClipper: videoClipperSlice,
    scenePicker: scenePickerSlice,
    captionScriptor: captionScriptorSlice,
    sceneDetail: sceneDetailSlice,
    audioLibrary: audioLibrarySlice,
    sourceScraper: sourceScraperSlice
  }
})

export default store
// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch
