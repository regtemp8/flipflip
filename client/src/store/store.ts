/// <reference path="../global.d.ts"/>
import { createReduxStore } from '../server/renderer'

const store = createReduxStore(
  window.flipflipAppStorage,
  window.flipflipConstants
)

export default store
// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch
