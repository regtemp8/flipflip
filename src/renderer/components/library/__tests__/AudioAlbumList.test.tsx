import React from "react";
import { describe, it, expect } from "@jest/globals";
import renderer from "react-test-renderer";
import AudioAlbumList from "../AudioAlbumList";
import TestProvider from "../../../../../test/util/TestProvider";
import store from "../../../../store/store";

describe("AudioAlbumList", () => {
  it("should match snapshot", () => {
    const component = renderer.create(
      <TestProvider store={store}>
        <AudioAlbumList
          sources={[]}
          showHelp={false}
          onClickAlbum={(album) => {}}
          onClickArtist={(artist) => {}}
        />
      </TestProvider>
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
