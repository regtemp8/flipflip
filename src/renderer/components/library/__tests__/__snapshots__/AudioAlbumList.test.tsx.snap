// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AudioAlbumList should match snapshot 1`] = `
[
  <h1
    className="MuiTypography-root MuiTypography-h3 MuiTypography-noWrap AudioAlbumList-emptyMessage-1 css-1yk4hom-MuiTypography-root"
  >
    乁( ◔ ౪◔)「
  </h1>,
  <h1
    className="MuiTypography-root MuiTypography-h4 MuiTypography-noWrap AudioAlbumList-emptyMessage2-2 css-1r3v6n7-MuiTypography-root"
  >
    Nothing here
  </h1>,
]
`;
